# Electronic Structure Theory
Our goal is to solve the time-indepdent Schrodinger's equation:
\begin{align}
\hat{H} |\psi \rangle = E |\psi \rangle
\end{align}

For molecular systems consisting of $N$ electrons and $M$ nuclei, the Hamiltonian in atomic units is given by:
\begin{align}
\hat{H} = -\sum_{i=1}^{N} \frac{1}{2} \nabla^2_i
          -\sum_{A=1}^{M} \frac{1}{2 M_A} \nabla^2_A
          -\sum_{i=1}^{N} \sum_{A=1}^{M} \frac{Z_A}{r_{iA}}
          +\sum_{i=1}^{N} \sum_{j>i}^{N} \frac{1}{r_{ij}}
          +\sum_{A=1}^{M} \sum_{B>A}^{M} \frac{Z_A Z_B}{R_{AB}} 
          ,
\end{align}
where $i$ and $j$ refer to electrons and $A$ and $B$ refer to nuclei.

The Hamiltonian contains kinetic energy terms for each electron and nucleus in the system. It also contains three potential energy terms: (a) attraction between the electrons and the nuclei; (b) repulsion between the nuclei; and (c) repulsion between the electrons. Thus, the Hamiltonian can be written as:
\begin{align}
\hat{H} = \hat{T}_N(\textbf{R}) 
          + \hat{T}_e(\textbf{r})
          + \hat{V}_{eN}(\textbf{r}, \textbf{R})
          + \hat{V}_{NN}(\textbf{R})
          + \hat{V}_{ee}(\textbf{r}),
\end{align}
where $\textbf{R}$ is the set of nuclear coordinates and $\textbf{r}$ is the set of electronic coordinates.

## The Born-Oppenheimer Approximation
Because of the interaction term between the electrons and nuclei $\hat{V}_{eN}(\textbf{r}, \textbf{R})$, the Hamiltonian and in turn the wavefunction cannot be exactly written as a product of a nuclear and electronic part (i.e., $\psi(\textbf{r}, \textbf{R}) = \psi_{e}(\textbf{r}) \chi (\textbf{R})$). However, nuclei are much heavier and slower than the electrons ($m_p \approx 1800m_e$). Thus, electrons can quickly relax for any given nuclear configuration. The separation of the wavefunction to nuclear and electronic parts is therefore *approximately* correct. This is the essence of the **Born-Oppenheimer Approximation**.

Now consider a given *fixed* nuclear configuration $\textbf{R}$. The $\hat{T}_N(\textbf{R})$ term will be zero and the $\hat{V}_{NN}(\textbf{R})$ term will be a constant. Additionally, the $\hat{V}_{eN}(\textbf{r}; \textbf{R})$ term will only have a *parameteric* dependence on $\textbf{R}$. Thus, we can define the electronic Hamiltonian as:
\begin{align}
\hat{H}_{e} = \hat{T}_e(\textbf{r})
              + \hat{V}_{eN}(\textbf{r}; \textbf{R})
              + \hat{V}_{ee}(\textbf{r})
\end{align}

Notice that $V_{NN}(\textbf{R})$ is often included in the electronic Hamiltonian. Given $\hat{H}_{e}$, we can define the electronic Schrodinger's equation:
\begin{align}
\hat{H}_{e} \psi_{e}(\textbf{r}; \textbf{R}) = E_{e} \psi_{e}(\textbf{r}; \textbf{R})
\end{align}

Solving this equation will yield the electronic energy $E_{e}$. Quantum chemistry is largely concerned with developing approximate methods for solving the electronic Schrodinger's equation.

Having solved the electronic Schrodinger's equation, we next proceed to finding the solution for the nuclear wavefunction $\chi(\textbf{R})$. Because the electrons move much faster than the nuclei, it is reasonable to assume that the nuclear motion will only experience the *average* effect of the electron interactions. Thus, we can define the nuclear Hamiltonian as follows:

\begin{align}
\hat{H}_{nucl} &= \hat{T}_N(\textbf{R}) +  \hat{V}_{NN}(\textbf{R}) + \langle \hat{H}_{e} \rangle \\
               &= \hat{T}_N(\textbf{R})  + \hat{V}_{NN}(\textbf{R}) + E_{e}
\end{align}

We therefore see that the nuclei move on a **potential energy surface** obtained by solving the electronic Schrodinger's equation.

Given this nuclear Hamiltonian, we can define a nuclear Schrodinger's equation:
\begin{align}
\hat{H}_{nucl} \chi = E \chi
\end{align}

Solving the nuclear Schrodinger's equation gives the energy $E$, which is the energy in the original Schrodinger's equation $\hat{H} |\psi \rangle = E |\psi \rangle$. It also yields the 
nuclear wavefunction $\chi$.

Thus, the Born-Oppenheimer approximation separates the electronic and nuclear degrees of freedom. $\psi_{e}(\textbf{r})$ describes the electronic energy levels for the molecule while $\chi (\textbf{R})$ describes the nuclear energy levels. The nuclear energy includes the contributions of vibration, rotation, and translation. Our focus will be on the electronic Schrodinger's equation.

## Electron Spin and Pauil Exclusion Principle
To completely describe the electron, it is not sufficient to describe only its spatial coordinate. The electron also has a **spin** coordinate that can have one of two values: $\pm 1/2$. The **Pauli exclusion principle** states that the wavefunction of fermionic systems (which have half-integer spins) must be *antisymmetric* with respect to the exchange any two particles. For electrons, this can be stated in a more familiar language as follows: No two electrons with the same spin can occupy the same orbital.

Mathematically, for nonrelativistic quantum mechanics, we introduce two spin functions $\alpha (\omega)$ and $\beta (\omega)$ that depend on an unspecified spin variable $\omega$. These two functions represent up and down or $\uparrow$ and $\downarrow$ spin states. In the bra-ket notation, the two spin states are represented by $|\alpha \rangle$ and $|\beta \rangle$. The spin states are orthonormal:

\begin{align}
\langle \alpha | \beta \rangle = \int d \omega \alpha^* (\omega) \beta (\omega) 
                               = \delta_{\alpha \beta}
\end{align}

In this formalism, the electron depends on both a spatial coordinate $\textbf{r}$ and a spin coordinate $\omega$. The four degrees of freedom of the electron are denoted $\textbf{x}$.
\begin{align}
\textbf{x} = \{\textbf{r}, \omega\}
\end{align}

For a system with $N$ electrons, the wavefunction depends on $N$ $\textbf{x}$ coordinates:
\begin{align}
\phi = \phi(\textbf{x}_1, \textbf{x}_2, \cdots, \textbf{x}_N)
\end{align}

With this wavefunction, the Pauli exclusion principle can be represented as follows:
\begin{align}
\phi(\textbf{x}_1, \cdots, \textbf{x}_i, \cdots, \textbf{x}_j, \cdots, \textbf{x}_N) = 
-\phi(\textbf{x}_1, \cdots, \textbf{x}_j, \cdots, \textbf{x}_i, \cdots, \textbf{x}_N).
\end{align}

That is, exchanging any two electrons $i$ and $j$ will result in changing sign of the wavefunction

## Spin and Spatial Orbitals
As was discussed for the hydrogen atom, the orbital amplitude $\psi^*\psi d\textbf{r}$ describes the probability of finding the electron in the volume element $d\textbf{r}$. For multi-electron systems, the orbital concept provides a very fruitful mechanism for approximating the wavefunction. As is known from introductory textbooks, chemists often construct **molecular orbital (MO)** from a **linear combination of atomic orbitals (LCAO)**.

Because the electron has both spatial and spin coordinates, the *spin orbital* $\chi(\textbf{x})$ will have a spatial and a spin component. We will represent the spatial component by $\psi(\textbf{r})$ and the spin component by either $\alpha(\omega)$ or $\beta(\omega)$. In other words,

\begin{align}
\chi(\textbf{x}) &= \psi(\textbf{r}) \alpha(\omega)
\end{align}

or

\begin{align}
\chi(\textbf{x}) &= \psi(\textbf{r}) \beta(\omega)
\end{align}

If the spatial orbitals $\psi_i$ form a complete set, then *any* spatial function can be written in terms of the $\psi_i$ **basis functions**:
\begin{align}
f(\textbf{r}) = \sum_{i=1}^\infty a_i \psi_i(\textbf{r}),
\end{align}
where $a_i$ are the expansion coefficients. In practice, the expansion is truncated to a finite number of the basis functions and the basis set is not complete.

## The Hartree Product
We will first discuss the form of the wavefunction for a two-electron system as an instructive case. Then, we will generalize the discussion to a general many-electron system.

Suppose the two electrons are completely indepedent and they do not have any interaction. The full Hamiltonian can be separated into two Hamiltonians: one for electron 1 and one for electron 2:

\begin{align}
\hat{H} = \hat{h}(1) + \hat{h}(2),
\end{align}

where $\hat{h}(i)$ is the **one-electron operator** for electron $i$. Furthermore, each electron will have its own set of eigenfunction or orbitals that can be found by solving:
\begin{align}
\hat{h}(1) \chi_i(\textbf{x}_1) &= \epsilon_i \chi_i(\textbf{x}_1) \\
\hat{h}(2) \chi_j(\textbf{x}_2) &= \epsilon_j \chi_j(\textbf{x}_2)
\end{align}

Because the Hamiltonian is separable, the total wavefunction will be a product of the wavefunctions of each electron:

\begin{align}
\Psi^{\rm HP}(\textbf{x}_1, \textbf{x}_2) = \chi_i(\textbf{x}_1) \chi_j(\textbf{x}_2)
\end{align}

and the total energy will be the sum of the energies:

\begin{align}
E = \epsilon_1 + \epsilon_2
\end{align}

This form of the wavefunction is called the **Hartree product** and it describes a truly indepdent particle model.

## The Slater Determinant
The main deficiency of the Hartree product wavefunction is that is does not satisfy the Pauli exclusion principle:
\begin{align}
\Psi^{\rm HP}(\textbf{x}_1, \textbf{x}_2) \neq -\Psi^{\rm HP}(\textbf{x}_2, \textbf{x}_1).
\end{align}

We can fix this deficiency by using a *linear combination* of Hartree products as our trial function:
\begin{align}
\Psi^{\rm SD}(\textbf{x}_1, \textbf{x}_2) = \frac{1}{\sqrt{2}} \left (
                                             \chi_i(\textbf{x}_1) \chi_j(\textbf{x}_2) -
                                             \chi_j(\textbf{x}_1) \chi_i(\textbf{x}_2)
                                             \right ),
\end{align}

where $\frac{1}{\sqrt{2}}$ is a normalization factor. With this wavefunction, we can see that it satisfies the Pauli exclusion principle:

\begin{align}
\Psi^{\rm SD}(\textbf{x}_1, \textbf{x}_2) = -\Psi^{\rm SD}(\textbf{x}_2, \textbf{x}_1)
\end{align}

We call this form of the wavefunction, the **Slater determinant**. Determinants have the important property that exchanging two columns or two rows changes the sign of the determinat. In the determinant notation, we can represent above wavefunction as:

\begin{align}
\Psi^{\rm SD}(\textbf{x}_1, \textbf{x}_2) = \frac{1}{\sqrt{2}}
    \begin{vmatrix}
         \chi_i(\textbf{x}_1) & \chi_j(\textbf{x}_1) \\
         \chi_i(\textbf{x}_2) & \chi_j(\textbf{x}_2)
    \end{vmatrix}
\end{align}

For a general $N$-electron system with $k$ spin orbitals, the **Slater determinant** becomes:
\begin{align}
\Psi^{\rm SD}(\textbf{x}_1, \textbf{x}_2, \cdots, \textbf{x}_N) = \frac{1}{\sqrt{N!}}
    \begin{vmatrix}
         \chi_i(\textbf{x}_1) & \chi_j(\textbf{x}_1) & \cdots & \chi_k(\textbf{x}_1) \\
         \chi_i(\textbf{x}_2) & \chi_j(\textbf{x}_2) & \cdots & \chi_k(\textbf{x}_2) \\
         \vdots               & \vdots               & \vdots & \vdots \\
         \chi_i(\textbf{x}_N) & \chi_j(\textbf{x}_N) & \vdots & \chi_k(\textbf{x}_N)
    \end{vmatrix}.
\end{align}

The rows of the determinant are labelled by the electrons ($\textbf{x}_1, \textbf{x}_2, \cdots, \textbf{x}_N$) and the columns are labelled by the spin orbitals ($\chi_i, \chi_j, \cdots, \chi_k$).


In [1]:
import sympy
from IPython.display import display

N = 3
chis = sympy.symbols("chi_i:%s" %str(chr(105+N)), cls=sympy.Function)
xs = sympy.symbols("x_1:%i" %(N+1))

matrix = sympy.zeros(N, N)
for i in range(N):
    for j in range(N):
        matrix[i, j] = chis[j](xs[i])

matrix *= 1/sympy.sqrt(sympy.factorial(N))

print("For %i electrons, the Slater determinant is given by the determinant of the matrix:" %N)
display(matrix)

print("which is equal to:")
SD = matrix.det()

display(SD)

print("We verify that the Slater determinant satisfies Pauli exclusion principle:")
d = sympy.symbols("d") # dummy variable
matrix2 = matrix.subs({xs[0]:d, xs[1]: xs[0]})
matrix2 = matrix2.subs({d: xs[1]})

print("Exchanging electrons 1 and 2, we get:")
display(matrix, matrix2)

print("The determinants are:")
display(matrix.det(), matrix2.det())

print("Are the determinants equal but with opposite signs?", matrix.det()==-matrix2.det())

print("Can electrons 1 and 2 occupy the same spin orbital?")
matrix3 = matrix.subs(chis[0], chis[1])
display(matrix3)

print("The value of the determinant is", matrix3.det())

For 3 electrons, the Slater determinant is given by the determinant of the matrix:


Matrix([
[sqrt(6)*chi_i(x_1)/6, sqrt(6)*chi_j(x_1)/6, sqrt(6)*chi_k(x_1)/6],
[sqrt(6)*chi_i(x_2)/6, sqrt(6)*chi_j(x_2)/6, sqrt(6)*chi_k(x_2)/6],
[sqrt(6)*chi_i(x_3)/6, sqrt(6)*chi_j(x_3)/6, sqrt(6)*chi_k(x_3)/6]])

which is equal to:


sqrt(6)*chi_i(x_1)*chi_j(x_2)*chi_k(x_3)/36 - sqrt(6)*chi_i(x_1)*chi_j(x_3)*chi_k(x_2)/36 - sqrt(6)*chi_i(x_2)*chi_j(x_1)*chi_k(x_3)/36 + sqrt(6)*chi_i(x_2)*chi_j(x_3)*chi_k(x_1)/36 + sqrt(6)*chi_i(x_3)*chi_j(x_1)*chi_k(x_2)/36 - sqrt(6)*chi_i(x_3)*chi_j(x_2)*chi_k(x_1)/36

We verify that the Slater determinant satisfies Pauli exclusion principle:
Exchanging electrons 1 and 2, we get:


Matrix([
[sqrt(6)*chi_i(x_1)/6, sqrt(6)*chi_j(x_1)/6, sqrt(6)*chi_k(x_1)/6],
[sqrt(6)*chi_i(x_2)/6, sqrt(6)*chi_j(x_2)/6, sqrt(6)*chi_k(x_2)/6],
[sqrt(6)*chi_i(x_3)/6, sqrt(6)*chi_j(x_3)/6, sqrt(6)*chi_k(x_3)/6]])

Matrix([
[sqrt(6)*chi_i(x_2)/6, sqrt(6)*chi_j(x_2)/6, sqrt(6)*chi_k(x_2)/6],
[sqrt(6)*chi_i(x_1)/6, sqrt(6)*chi_j(x_1)/6, sqrt(6)*chi_k(x_1)/6],
[sqrt(6)*chi_i(x_3)/6, sqrt(6)*chi_j(x_3)/6, sqrt(6)*chi_k(x_3)/6]])

The determinants are:


sqrt(6)*chi_i(x_1)*chi_j(x_2)*chi_k(x_3)/36 - sqrt(6)*chi_i(x_1)*chi_j(x_3)*chi_k(x_2)/36 - sqrt(6)*chi_i(x_2)*chi_j(x_1)*chi_k(x_3)/36 + sqrt(6)*chi_i(x_2)*chi_j(x_3)*chi_k(x_1)/36 + sqrt(6)*chi_i(x_3)*chi_j(x_1)*chi_k(x_2)/36 - sqrt(6)*chi_i(x_3)*chi_j(x_2)*chi_k(x_1)/36

-sqrt(6)*chi_i(x_1)*chi_j(x_2)*chi_k(x_3)/36 + sqrt(6)*chi_i(x_1)*chi_j(x_3)*chi_k(x_2)/36 + sqrt(6)*chi_i(x_2)*chi_j(x_1)*chi_k(x_3)/36 - sqrt(6)*chi_i(x_2)*chi_j(x_3)*chi_k(x_1)/36 - sqrt(6)*chi_i(x_3)*chi_j(x_1)*chi_k(x_2)/36 + sqrt(6)*chi_i(x_3)*chi_j(x_2)*chi_k(x_1)/36

Are the determinants equal but with opposite signs? True
Can electrons 1 and 2 occupy the same spin orbital?


Matrix([
[sqrt(6)*chi_j(x_1)/6, sqrt(6)*chi_j(x_1)/6, sqrt(6)*chi_k(x_1)/6],
[sqrt(6)*chi_j(x_2)/6, sqrt(6)*chi_j(x_2)/6, sqrt(6)*chi_k(x_2)/6],
[sqrt(6)*chi_j(x_3)/6, sqrt(6)*chi_j(x_3)/6, sqrt(6)*chi_k(x_3)/6]])

The value of the determinant is 0


Clearly, writing down all the terms in the Slater determinant is very tedious. So, it is useful to introduce the shorthand notation:
\begin{align}
\Psi^{\rm SD}(\textbf{x}_1, \textbf{x}_2, \cdots, \textbf{x}_N) = 
    |\chi_i(\textbf{x}_1) \chi_j(\textbf{x}_2) \cdots \chi_k(\textbf{x}_N) \rangle.
\end{align}

This notation for a *normalized* wavefunction shows only the diagonal elements in the determinant and suppresses the normalization constant.

If we choose the electron labels to be always in the order $\textbf{x}_1, \textbf{x}_2, \cdots, \textbf{x}_N$, then we can use the simpler notation:
\begin{align}
\Psi^{\rm SD}(\textbf{x}_1, \textbf{x}_2, \cdots, \textbf{x}_N) = 
    |\chi_i \chi_j \cdots \chi_k \rangle
\end{align}
or even
\begin{align}
\Psi^{\rm SD}(\textbf{x}_1, \textbf{x}_2, \cdots, \textbf{x}_N) = |i j \cdots k \rangle.
\end{align}

### Exchange Correlation
Both the Hartree product and the Slater determinants are independent particle models because they write the wavefunction in terms of the one-electron spin orbitals $\chi$. However, antisymmetrizing the wavefunction in the Slater determinant introduces **exchange** correlation effects. To understand this, consider the parobabilty amplitude for the Hartree product of two electrons:

\begin{align}
|\Psi^{\rm HP}(\textbf{x}_1, \textbf{x}_2)|^2 d\textbf{x}_1 d\textbf{x}_2
 = |\chi_i(\textbf{x}_1)|^2 d\textbf{x}_1 |\chi_j(\textbf{x}_2)|^2 d\textbf{x}_2
\end{align}

This is just the product of the probability amplitudes for each electron. So it is possible that two electrons with the same spin occupy the same point in space.

Now consider the Slater determinant for two electrons that have opposite spins:
\begin{align}
\Psi^{\rm SD}_{\uparrow \downarrow}(\textbf{x}_1, \textbf{x}_2) &= |\chi_i \chi_j \rangle \\
\chi_i(\textbf{x}_1) &= \psi_i(\textbf{r}_1) \alpha(\omega_1) \\
\chi_j(\textbf{x}_2) &= \psi_j(\textbf{r}_2) \beta(\omega_2)
\end{align}

The probability amplitude becomes:
\begin{align}
|\Psi^{\rm SD}_{\uparrow \downarrow}(\textbf{x}_1, \textbf{x}_2)|^2 d\textbf{x}_1 d\textbf{x}_2 =
\frac{1}{2} |\psi_i(\textbf{r}_1) \alpha(\omega_1) \psi_j(\textbf{r}_2) \beta(\omega_2)
-\psi_i(\textbf{r}_2) \alpha(\omega_2) \psi_j(\textbf{r}_1) \beta(\omega_1)
|^2
\end{align}

Integrating (averaging) out the spin coordinates, we obtain the probability of finding electron 1 in $\textbf{r}_1$ and electron 2 in $\textbf{r}_2$:
\begin{align}
P(\textbf{r}_1, \textbf{r}_2) d\textbf{r}_1 d\textbf{r}_2 &=
\int d\omega_1 d\omega_2 
|\Psi^{\rm SD}_{\uparrow \downarrow}|^2 d\textbf{r}_1 d\textbf{r}_2  \\
&= \frac{1}{2} \left (  |\psi_i(\textbf{r}_1)|^2 |\psi_j(\textbf{r}_2)|^2                    
+ |\psi_i(\textbf{r}_2)|^2 |\psi_j(\textbf{r}_1)|^2
\right )
d\textbf{r}_1 d\textbf{r}_2
\end{align}

The fisrt term is the probability of finding electron 1 at $d\textbf{r}_1$ and electron 2 at $d\textbf{r}_2$. The second term is the probability of finding electron 1 at $d\textbf{r}_2$ and electron 2 at $d\textbf{r}_1$. Because electrons are indistinguishable, the correct probability is the average of the two probabilities. However, the motion of the two electrons are uncorrelated and they can occupy the same position in space ($P(\textbf{r}_1, \textbf{r}_1) \neq 0$ if $\psi_i = \psi_j$).

Now consider a similar analysis for two electrons with the same spin:
\begin{align}
\chi_i(\textbf{x}_1) &= \psi_i(\textbf{r}_1) \beta(\omega_1) \\
\chi_j(\textbf{x}_2) &= \psi_j(\textbf{r}_2) \beta(\omega_2)
\end{align}

The probability becomes:
\begin{align}
P(\textbf{r}_1, \textbf{r}_2) = \frac{1}{2} \left \{ 
|\psi_i(\textbf{r}_1)|^2 |\psi_j(\textbf{r}_2)|^2                    
+ |\psi_i(\textbf{r}_2)|^2 |\psi_j(\textbf{r}_1)|^2
\\ - \left [
\psi_i^*(\textbf{r}_1) \psi_j(\textbf{r}_1) \psi_j^*(\textbf{r}_2) \psi_i(\textbf{r}_2)
+ \psi_i(\textbf{r}_1) \psi_j^*(\textbf{r}_1) \psi_j(\textbf{r}_2) \psi_i^*(\textbf{r}_2)
\right ]
\right \}
\end{align}

The extra corss term makes the probability correlated for electrons with the same spin. In particular, if electrons 1 and 2 occupy the same spatial orbital $\psi$, then $P(\textbf{r}_1, \textbf{r}_1)=0$ and no electrons with the same spin can occupy the same orbital.

### Visual Representation of Slater Determinants for the H$_2$ Minimal Basis Set Model
We can use a basis set of spin orbitals to represent the wavefunction. If our basis set is complete, then we will have an infinite number of orbitals. In practice, we use a finite number of basis functions. If we have $K$ spatial orbitals $\psi$, then we will have $2K$ spin orbitals $\chi$ because we have two spin states $\alpha$ and $\beta$. In general, some of these orbials will be *occupied* and some will empty, *unocuppied* or *virtual*.

We will illustrate these concepts for H$_2$ using a *minimal basis set* model. That is, we will represent each hydrogen atom by a single orbital $\phi$. Then, we will construct *molecular orbitals* using *linear combination of atomic orbitals*.

The normalized spatial molecular orbitals for H$_2$ will be given by:
\begin{align}
\psi_1 &= \frac{1}{\sqrt{2(1+S_{12})}} \left ( \phi_1 + \phi_2 \right ) \quad (\textrm{Bonding orbital}) \\
\psi_2 &= \frac{1}{\sqrt{2(1-S_{12})}} \left (\phi_1 - \phi_2 \right ) \quad (\textrm{Antibonding orbital}) \\
S_{12} &= \int d\textbf{r} \phi_1^*(\textbf{r}) \phi_2(\textbf{r})
\end{align}

From these two spatial orbitals, we can construct four spin orbitals:
\begin{align}
\chi_1(\textbf{x}) &= \psi_1(\textbf{r}) \alpha(\omega) \\
\chi_2(\textbf{x}) &= \psi_1(\textbf{r}) \beta(\omega) \\
\chi_3(\textbf{x}) &= \psi_2(\textbf{r}) \alpha(\omega) \\
\chi_4(\textbf{x}) &= \psi_2(\textbf{r}) \beta(\omega) \\
\end{align}

Because H$_2$ has two electrons, only two of the spin orbitals will be occupied. The other two will be unoccupied. For the ground state, we can represent this visually as follows:
\begin{align}
|\Psi_0 \rangle = &\, \chi_3 \,\underline{\:\:}   \,   \underline{\:\:}\,\, \chi_4 \\
                  &\, \chi_1 \underline{\uparrow} \,   \underline{\downarrow} \chi_2 \\
                = & \underline{\quad} \, \psi_2 \\
                  & \underline{\uparrow \downarrow} \, \psi_1.
\end{align}
Mathematically,
\begin{align}
|\Psi_0 \rangle = |\chi_1 \chi_2 \rangle = |\psi_1 \overline{\psi}_1 \rangle = | 1 \overline{1} \rangle,
\end{align}

where the absence and presence of the bar indicates the two different spin states.

For the hydrogen molecule in a minimal basis set, there are four configurations with singly excited electron:
\begin{align}
| \Psi_1^2 \rangle = |2\overline{1} \rangle = & \underline{\uparrow}  \, \psi_2\\
                                              & \underline{\downarrow} \, \psi_1,
\end{align}

\begin{align}
|\Psi_1^\overline{2} \rangle = |\overline{21} \rangle = & \underline{\downarrow}  \, \psi_2\\
                                                        & \underline{\downarrow}  \, \psi_1,
\end{align}

\begin{align}
|\Psi_\overline{1}^2 \rangle = |12 \rangle = & \underline{\uparrow}  \, \psi_2\\
                                                        & \underline{\uparrow}  \, \psi_1,
\end{align}
and

\begin{align}
|\Psi_\overline{1}^\overline{2} \rangle = |1\overline{2} \rangle = & \underline{\downarrow} \psi_2\\
                                                        & \underline{\uparrow}  \, \psi_1.
\end{align}

Finally, there is a single state with doubly excited electrons:
\begin{align}
|\Psi_{1 \overline{1}}^{2 \overline{2}} \rangle = |2 \overline{2} \rangle = 
    & \underline{\uparrow \downarrow} \, \psi_2 \\
    & \underline{\quad} \, \psi_1. 
\end{align}

## The Hartree-Fock Theory

### The one- and two-electron integrals
The Hartree-Fock theory is a procedure for calculating the best energy, *in a variational sense*, using a single Slater determinant $|\Psi_0 \rangle$ as the trial wavefunction. It turns out that using a single Slater determinant implies that each electron interacts with the *average* electron density of all other electrons. This is why the Hartree-Fock theory is considered a *mean-field theory*.

Let's first define a simpler notation for the electronic Hamiltonian for many-electron systems:
\begin{align}
\hat{H}_{e} & = \hat{T}_e(\textbf{r})
              + \hat{V}_{eN}(\textbf{r}; \textbf{R})
              + \hat{V}_{ee}(\textbf{r}) \\
            & = -\sum_{i=1}^{N} \frac{1}{2} \nabla^2_i
                -\sum_{i=1}^{N} \sum_{A=1}^{M} \frac{Z_A}{r_{iA}}
                +\sum_{i=1}^{N} \sum_{j>i}^{N} \frac{1}{r_{ij}} \\
            & = \sum_i \hat{h}(i) + \sum_{j>i} \hat{v}(i, j).
\end{align}

$\hat{h}(i)$ is again the **one-electron operator** and it consists of the first two terms. $\hat{v}(i, j)$ is the **two-electron operator**. To calculate the average energy, we use the usual formula:
\begin{align}
E = \langle \Psi| \hat{H}_{e} | \Psi \rangle = 
\sum_i \langle \Psi| \hat{h}(i) | \Psi \rangle +
\sum_{j>i} \langle \Psi| \hat{v}(i, j) | \Psi \rangle
\end{align}

We therefore need to to calculate the integrals for $\hat{h}$ and for $\hat{v}$.

The one-electron integral is given by:
\begin{align}
[i|h|j] = \int d\textbf{x}_1 \chi_i^*(\textbf{x}_1) h(\textbf{r}_1) \chi_j(\textbf{x}_1).
\end{align}

For the two-electron integrals:
\begin{align}
[ij|kl] = \int d\textbf{x}_1 d\textbf{x}_2 \chi_i^*(\textbf{x}_1) \chi_j(\textbf{x}_1)
r_{12}^{-1} \chi_k^*(\textbf{x}_2) \chi_l(\textbf{x}_2)
\end{align}

The integrations are over the spin-spatial coordinate ($\textbf{x}$). This is called the *chemists' notation*, which puts the orbitals of electron 1 in the left and the two orbitals of electron 2 in the right. Another commonly used notation is the *physicists' notation*:
\begin{align}
\langle ij|kl \rangle = \int d\textbf{x}_1 d\textbf{x}_2 \chi_i^*(\textbf{x}_1) \chi_j^*(\textbf{x}_2)
r_{12}^{-1} \chi_k(\textbf{x}_1) \chi_l(\textbf{x}_2)
\end{align}
where both complex conjugated orbitals of electrons 1 and 2 are on the left. The conversion from the chemists' to the physicists notations is given by:
\begin{align}
[ij|kl] = \langle ik|jl \rangle.
\end{align}

In terms of the spatial coordinate $\textbf{r}$ only, the integrals in the chemists' notations are given by:
\begin{align}
(i|h|j) &= h_{ij} = (\psi_i|h|\psi_j) = \int d\textbf{r}_1 \psi_i^*(\textbf{r}_1) h(\textbf{r}_1) \psi_j(\textbf{r}_1) \\
(ij|kl) &= (\psi_i \psi_j | \psi_k \psi_l ) = \int d\textbf{r}_1 d\textbf{r}_2 \psi_i^*(\textbf{r}_1) \psi_j(\textbf{r}_1)
r_{12}^{-1} \psi_k^*(\textbf{r}_2) \psi_l(\textbf{r}_2)
\end{align}

In Hartree-Fock theory, only two terms are relavent for the two-electron integrals:
\begin{align}
J_{ij} &= (ii | jj) \quad \textrm{Coulomb integrals} \\
K_{ij} &= (ij | ji) \quad \textrm{Exchange integrals}
\end{align}

For exchange integrals, the spins for electrons $i$ and $j$ have to be identical. Otherwise, the exchange term will be zero.

### The Hartree-Fock Energy
In the Hartree-Fock procedue, we use a single Slater determinant to represent the trial wavefunction. Suppose we have the Slater determinant $|\Psi_0 \rangle = |ab \cdots \rangle$, where $a, b$, etc, are the occupied orbitals. If we substitute this wavefunction in the energy expectation value expression, we obtain the Hartree-Fock energy of the system in terms of the one- and two-electron integrals:

\begin{align}
E_{HF} &= \sum_i  [i|h|i] + \sum_{i>j} [ii|jj] - [ij|ji],
\end{align}
where the sum is over all the occupied spin orbitals. $[ii|jj]$ is the Coulumb integral and $[ij|ji]$ is the exchange integral.

For example, consider the helium atom:
\begin{align}
|\Psi_0 \rangle = |\chi_1 \chi_2 \rangle = |1 2 \rangle = \underline{\uparrow} \, \underline{\downarrow}
\end{align}

The energy will be:
\begin{align}
E_{HF} = [1|h|1] + [2|h|2] + [22|11] - [21|12], 
\end{align}
where $1$ and $2$ are the indices of the spin orbitals $\chi$. However, the last term is in fact equal to zero. Remember that only electrons that have the same spin have exchange correlation. Mathematically, we can show that the last term is equal to zero by integrating the spin coordinates. Suppose first that:
\begin{align}
\chi_1 &= \psi_1 \alpha \\
\chi_2 &= \psi_2 \beta \\
\end{align}

Then, the integral becomes:
\begin{align}
[21|12] &= \int d\textbf{x}_1 d\textbf{x}_2 \chi_2^*(\textbf{x}_1) \chi_1(\textbf{x}_1)
r_{12}^{-1} \chi_1^*(\textbf{x}_2) \chi_2(\textbf{x}_2) \\
        &= \int \omega_1 \int \omega_2 \int d\textbf{r}_1 \int d\textbf{r}_2
        \psi_2^*(\textbf{r}_1) \beta^*(\omega_1)  \psi_1(\textbf{r}_1) \alpha(\omega_1)
        r_{12}^{-1}
        \psi_1^*(\textbf{r}_2) \alpha^*(\omega_2)  \psi_2(\textbf{r}_2) \beta(\omega_2) \\
        &= \int d \omega_1 \beta^*(\omega_1) \alpha(\omega_1)
           \int d \omega_2 \alpha^*(\omega_2) \beta(\omega_2)
           \, (21|12).
\end{align}

The spin integrations are equal to zero.

### Restricted and Unrestricted Hartree-Fock Theory
There are two common flavors of Hartree-Fock theory depending on how we treat the spin and the spatial components of the spin orbitals. Consider again the helium atom:
\begin{align}
|\Psi_0 \rangle &= |\chi_1 \chi_2 \rangle = |1 2 \rangle = \underline{\uparrow} \, \underline{\downarrow} \\
\chi_1 &= \psi_1 \alpha \\
\chi_2 &= \psi_2 \beta 
\end{align}

In **restricted Hartree-Fock (RHF)** theory, we assume the spatial orbitals $\psi_1$ and $\psi_2$ are the same. In **unrestricted Hartree-Fock (UHF)** theory, we assume the spatial orbitals $\psi_1$ and $\psi_2$ are different. Restricted Hartree-Fock theory is only applicable for *closed-shell* systems, that is systems with all electrons paired. Unrestricted Hartree-Fock theory is applicable for both *closed-shell* and *open-shell* systems, those which have unpaired electrons. The problem with the UHF method is that the wavefunction is not an eigenfunction of the spin operator $\textbf{S}^2$. For example, a "singlet" UHF wave function may also contain contributions from higher-lying triplet,
quintet, etc. This is called **spin-contamination**.

Consider now the Li atom. The electronic configuration is $1s^2 2s^1$, so the valence electron is unpaired. The wavefunction is given by:
\begin{align}
|\Psi_0 \rangle = |\chi_1 \chi_2 \chi_3 \rangle = |1 2 3\rangle = &\underline{\uparrow} \\
&\underline{\uparrow} \, \underline{\downarrow}
\end{align}

The spin orbitals are:
\begin{align}
\chi_1 &= \psi_1 \alpha \\
\chi_2 &= \psi_2 \beta \\
\chi_3 &= \psi_3 \alpha 
\end{align}

Clearly, restricted Hartree-Fock theory is inappropriate for this system. However, we have two options for treating the *core* paired electrons. If we assume that they occupy the same spatial orbital (i.e., $\psi_1=\psi_2$), then we can have a third flavor of Hartree-Fock theory called **restricted open-shell Hartree-Fock (ROHF)**. If we assume $\psi_1 \neq \psi_2$, then we can use unrestricted Hartree-Fock theory.

Consider again the helium atom:
\begin{align}
|\Psi_0 \rangle = |\chi_1 \chi_2 \rangle = |1 2 \rangle = \underline{\uparrow} \, \underline{\downarrow}
\end{align}

The energy will be:
\begin{align}
E_{HF} = [1|h|1] + [2|h|2] + [22|11], 
\end{align}
where $1$ and $2$ are the indices of the spin orbitals $\chi$. We can also write the energy as:
\begin{align}
E_{HF} = [1|h|1] + [\overline{1}|h|\overline{1}] + [\overline{11}|11],
\end{align}
where again the bar indicates the opposite spin. Integrating out the spins, we can get the energy in terms of the spatial orbitals:
\begin{align}
E_{HF} = (1|h|1) + (\overline{1}|h|\overline{1}) + (\overline{11}|11).
\end{align}

For restricted Hartree-Fock theory, the spatial orbitals are the same and thus we get:
\begin{align}
E_{RHF} = 2 (1|h|1) + (11|11). 
\end{align}

Now let's write the energy for the Li atom:
\begin{align}
\underset{1s}{\underline{\uparrow \downarrow}} \,  \underset{2s}{\underline{\uparrow \phantom{\downarrow}}}
\end{align}
* The three electrons will contribute 3 one-electron integral terms
* The three pairs of electrons will contribute three Coulomb terms
* The two electrons with the same spin will contribute one exchange term

Thus, the UHF energy will be:
\begin{align}
E_{UHF} &= (1|h|1) + (\overline{1}|h|\overline{1}) + (2|h|2) + 
         (\overline{11}|11) + (11|22) + (\overline{11}|22) -
         (21|12) \\
       &= h_{11} + h_{\overline{11}} + h_{22} + 
          J_{\overline{1}1} + J_{21} + J_{2\overline{1}} -
          K_{21}
\end{align}

### Examples on Spin and Multiplicity

Below are the electronic configurations for the first ten atoms in the periodic table:

H: 1s$^1$   
He: 1s$^2$   
Li: 1s$^2$ 2s$^1$  
Be: 1s$^2$ 2s$^2$  
B: 1s$^2$ 2s$^2$ 2p$^1$  
C: 1s$^2$ 2s$^2$ 2p$^2$  
N: 1s$^2$ 2s$^2$ 2p$^3$  
O: 1s$^2$ 2s$^2$ 2p$^4$  
F: 1s$^2$ 2s$^2$ 2p$^5$  
Ne: 1s$^2$ 2s$^2$ 2p$^6$  

The above electronic configurations do not indicate the spin state of the electrons. **Hund's rule** dictates that high spin states are favored over low spin states for the same electronic configuration. Thus, we can draw the electron configuration for the last subshell as follows:

H:  $\underset{1s}{\underline{\uparrow}}$  
He: $\underset{1s}{\underline{\uparrow \downarrow}}$  
Li: $\underset{2s}{\underline{\uparrow}}$  
Be: $\underset{2s}{\underline{\uparrow \downarrow}}$  
B:  $\underset{2p_1}{\underline{\uparrow}} \, \underset{2p_2}{\underline{\phantom{\uparrow}}} \, \underset{2p_3}{\underline{\phantom{\uparrow}}}$  
C: $\underset{2p_1}{\underline{\uparrow}} \, \underset{2p_2}{\underline{\uparrow}} \, \underset{2p_3}{\underline{\phantom{\uparrow}}}$  
N: $\underset{2p_1}{\underline{\uparrow}} \, \underset{2p_2}{\underline{\uparrow}} \, \underset{2p_3}{\underline{\uparrow}}$  
O: $\underset{2p_1}{\underline{\uparrow \downarrow}} \, \underset{2p_2}{\underline{\uparrow}} \, \underset{2p_3}{\underline{\uparrow}}$  
F: $\underset{2p_1}{\underline{\uparrow \downarrow}} \, \underset{2p_2}{\underline{\uparrow \downarrow}} \, \underset{2p_3}{\underline{\uparrow}}$  
Ne: $\underset{2p_1}{\underline{\uparrow \downarrow}} \, \underset{2p_2}{\underline{\uparrow \downarrow}} \, \underset{2p_3}{\underline{\uparrow \downarrow}}$  

The spin **multiplicity** is equal to the number of unpaired electrons + 1. Thus, the multiplicities of the first ten atoms are: 

|Element | Multiplicity|
|--------|-------------|
|H       |2            |
|He      |1            |
|Li      |2            |
|Be      |1            |
|B       |2            |
|C       |3            |
|N       |4            |
|O       |3            |
|F       |2            |
|Ne      |1            |

Atoms or molecules that have unpaired electrons are called **open-shell** whereas those that do not have unpaired electrons are called **closed-shell**. Multiplicities of 1, 2, 3, etc, are called singlet, doublet, triplet, etc.

Given the above table, we can determine which of the Hartree-Fock spin formulations can be applied to which atom:

|Element |RHF |UHF |ROHF |
|--------|----|----|-----|
|H       |N   |Y   |Y    |
|He      |Y   |Y   |Y    |
|Li      |N   |Y   |Y    |     
|Be      |Y   |Y   |Y    |     
|B       |N   |Y   |Y    |
|C       |N   |Y   |Y    |
|N       |N   |Y   |Y    |
|O       |N   |Y   |Y    |
|F       |N   |Y   |Y    |
|Ne      |Y   |Y   |Y    |

We show below an example on Hartree-Fock spin formulations using the Psi4 quantum chemistry package. We will calculate the energy using Hartree-Fock theory and the Aug-cc-pVQZ basis set. We will see discuss later the meaning of basis sets 

In [6]:
import psi4

psi4.core.set_output_file('output.dat', True)
psi4.core.clean_options()

# For Hydrogen atom, the rhf option will raise an error because hydrogen atom is not a singlet
psi4.set_options({"scf_type": "pk",
                  "reference": "rhf"})

# The first line defines the charge and multiplicity
psi4.geometry("""0 2
H
""")

e = psi4.energy("scf/Aug-cc-pVQZ")
print("The energy is", e)

RuntimeError: 
Fatal Error: RHF: RHF reference is only for singlets.
Error occurred in file: /scratch/psilocaluser/conda-builds/psi4-multiout_1638121835567/work/psi4/src/psi4/libscf_solver/rhf.cc on line: 94
The most recent 5 function calls were:



In [5]:
import psi4

psi4.core.set_output_file('output.dat', True)
psi4.core.clean_options()

# For Hydrogen atom, the rhf error will raise an error because hydrogen atom is not a singlet
psi4.set_options({"scf_type": "pk",
                  "reference": "uhf"})

# The first line defines the charge and multiplicity
psi4.geometry("""0 2
H
""")

e = psi4.energy("scf/Aug-cc-pVQZ")
print("The energy is", e)

The energy is -0.49994832146911894


We will discuss below how Hartree-Fock calculations can be performed. However, we will illustrate here how the relaxing the restriction on the spatial orbitals in unrestricted Hartree-Fock theory affects the energy. The following table shows the Hartree-Fock energy computed using the aug-cc-pVDZ basis set for the first ten atoms and using different Hartree-Fock spin formulations:

|Element |RHF         |UHF         |ROHF         |
|--------|------------|------------|-------------|
|H       |N/A         | -0.499948  | -0.499948   |
|He      |-2.861522   | -2.861522  |-2.861522    |
|Li      |N/A         | -7.432719  | -7.432695   |
|Be      |-14.572969  |-14.572969  |-14.572969   |
|B       | N/A        | -24.532984 | -24.528975  |
|C       | N/A        | -37.693352 | -37.688323  |
|N       | N/A        | -54.403820 | -54.400225  |
|O       | N/A        | -74.817625 | -74.811064  |
|F       | N/A        | -99.414085 | -99.409209  |
|Ne      | -128.543756| -128.543756| -128.543756 |

The first observation that we can make is that for closed-shell atoms, the results of the three methods are identical. Relaxing the spatial constraint on the orbitals in UHF does not lead to a different solution from the restricted case. This is also generally true for molecules in their equilibrium geometries. However, in some cases, using UHF instead of RHF can lead to a better approximation of the energy, such as in bond dissociation. 

The second observation is that the UHF energy is lower than or equal to the ROHF energy. Why is this the case? It is because of the variational principle. The Hartree-Fock method is variational. Thus, the relaxation of the spatial constraint introduces additional variational freedom that lowers the energy. This, however, happens at the expense of spin contamination. Minimizing the energy does not gurantee that other properties derived from the wave function are accurate.

### Demonstrating Hund's Rule
We used above Hund's rule, which dictates that high spin states are favored over low spin states for the same electronic configuration. We will demonstrate here why this is the case. Consider the following two configurations for the electrons in the $p$ orbitals for carbon:
\begin{align}
1:& \underset{1}{\underline{\uparrow}} \, \underset{2}{\underline{\uparrow}} \, \underset{3}{\underline{\phantom{\uparrow}}} \\
2:& \underset{1}{\underline{\uparrow \downarrow}} \, \underset{2}{\underline{\phantom{\uparrow}}} \, \underset{3}{\underline{\phantom{\uparrow}}}
\end{align}

The Hartree-Fock energies of the two configurations using the one-electron integral $h$ and the coulomb ($J$) and exchange ($K$) integrals are given by:

- $E_{HF} = h_1 + h_2 + J_{12} - K_{12}$
- $E_{HF} = 2 h_1 + J_{11}$

The first energy is expected to be lower because of the substraction of the exchange term. We will verify by calculation that the high spin state is lower in energy.

In [4]:
import psi4

psi4.core.clean_options()
psi4.core.set_output_file('output.dat', True)

psi4.set_options({"scf_type": "pk", "reference": "uhf"})

psi4.geometry("""0 3
C""")
high_spin_energy = psi4.energy("scf/Aug-cc-pVQZ")

psi4.geometry("""0 1
C""")
low_spin_energy = psi4.energy("scf/Aug-cc-pVQZ")

print("The high spin energy is", high_spin_energy)
print("The low spin energy is", low_spin_energy)
print("The difference is", high_spin_energy - low_spin_energy)

The high spin energy is -37.693351536099286
The low spin energy is -37.604542647712826
The difference is -0.08880888838645973


Now, we do a similar calculation for oxygen.

In [7]:
import psi4

psi4.core.clean_options()
psi4.core.set_output_file('output.dat', True)

psi4.set_options({"scf_type": "pk", "reference": "uhf"})

psi4.geometry("""0 3
O""")
high_spin_energy = psi4.energy("scf/Aug-cc-pVQZ")

psi4.geometry("""0 1
O""")
low_spin_energy = psi4.energy("scf/Aug-cc-pVQZ")

print("The high spin energy is", high_spin_energy)
print("The low spin energy is", low_spin_energy)
print("The difference is", high_spin_energy - low_spin_energy)

The high spin energy is -74.81762505826256
The low spin energy is -74.68999499161824
The difference is -0.12763006664432908


### Example on Calculating the Hartree-Fock Energy 
We will calculate the energy of the helium atom in a minimal basis set as an example. Again, the wavefunction for helium is:
\begin{align}
|\Psi_0 \rangle = |\chi_1 \chi_2 \rangle = |1 2 \rangle = \underline{\uparrow} \, \underline{\downarrow}
\end{align}

The helium ground state is a singlet state with two paired electrons. For the spatial wavefunction, we will use the hydrogenic $1s$ orbital for the two electrons. The total energy of the helium atom will be:

\begin{align}
E_{RHF} = 2 (1|h|1) + (11|11). 
\end{align}

Notice that we do not need to perform any calculations to get the $(1|h|1)$ term. The hydrogenic $1s$ orbital is an eigenfunction of the one-electron operator, so we already know the energy. We only need to calculate the two-electron integral.

For reference, the experimental value for the ground state of helium is -2.9 Hartree.

In [6]:
from IPython.display import display
import sympy.physics.hydrogen

# Get the hydrogenic 1s orbital for helium 
r, theta, phi = sympy.symbols("r, theta, phi", real=True, positive=True)
psi100 = sympy.physics.hydrogen.Psi_nlm(1, 0, 0, r, phi, theta, Z=2)

# Get the 1-electron energy for helium
one_electron_energy = 2*sympy.physics.hydrogen.E_nl(1, Z=2)

print("The 1s orbital for helium is:")
display(psi100)

print("The one-electron energy for helium is:")
print(one_electron_energy)

The 1s orbital for helium is:


2*sqrt(2)*exp(-2*r)/sqrt(pi)

The 1-electron energy for helium is:
-4


Clearly, if we neglect electron-electron repulsion completely, we get a significant error of -1.1 Hartrees. The electrons are attracted by too much. Obviouslly, electron-electron repulsion will decrease this favorable attraction. Therefore, we can improve the energy by calculating the two electron integral.

In [7]:
# This example is from Griffiths Introduction to Quantum Mechanics
# Now we need to perform the 2-electron integral
# To carry out the integral, we will set the polar axis along r1 so that
# |r1-r2| = sqrt(r1^2 + r2^2 - 2 r1 r2 cos(theta2))
# Additionally, we need to split the r2 integral into two pieces:
# one ranging from 0 to r1 and the other from r1 to infinity

import scipy
import scipy.integrate
import numpy as np

def psi1(r):
    # This is the hydrogenic 1s orbital for helium
    return 2*np.sqrt(2/np.pi)*np.exp(-2*r)

def psi2(r):
    # We will use the same orbital for both electrons
    return psi1(r)

def r12(theta2, r2, r1):
    # Distance between the two electrons
    angle = np.cos(theta2)
    return np.sqrt(r1**2 + r2**2 - 2*r1*r2*angle)

def two_electron_e(theta2, r2, r1):
    # The volume element in spherical coordinate is dV=r^2 sin(theta)dr dtheta dphi
    dV1 = 4*np.pi*r1**2
    dV2 = 2*np.pi*r2**2*np.sin(theta2)
    distance = r12(theta2, r2, r1)

    return dV1*dV2*psi1(r1)*psi2(r1)/distance*psi1(r2)*psi2(r2)

def r1_bound():
    return (0, np.inf)

def r2_bound_1(r1):
    return (0, r1)

def r2_bound_2(r1):
    return (r1, np.inf)

repulsion = scipy.integrate.nquad(two_electron_e, [[0, np.pi], r2_bound_1, r1_bound])[0]
repulsion += scipy.integrate.nquad(two_electron_e, [[0, np.pi], r2_bound_2, r1_bound])[0]

total = one_electron_energy + repulsion
print("The two-electron Couloumb repulsion is", repulsion)
print("The total energy is", total)
print("The error is", total-(-2.9))

The two-electron Couloumb repulsion is 1.250000000004479
The total energy is -2.74999999999552
The error is 0.150000000004479


### The Hartree-Fock Equations
Again, the variational theorem states that for any trial wavefunction $\psi$,
\begin{align}
E_{gs} \le E_{trial} \equiv \langle\Psi|\hat{H}|\Psi \rangle.
\end{align}

Thus, we need to minimize the Hartree-Fock energy with respect to the orbitals. If we do this, we get the equation:

\begin{align}
h(\textbf{x}_1) \chi(\textbf{x}_1) + \sum_{j \neq i} \left [ \int d\textbf{x}_2 | \chi_j(\textbf{x}_2)|^2 r_{12}^{-1} \right ] \chi_i (\textbf{x}_1) - \sum_{j \neq i} \left [ \int d\textbf{x}_2 \chi_j^*(\textbf{x}_2) \chi_i (\textbf{x}_2) r_{12}^{-1} \right]\chi_j(\textbf{x}_1) = \sum_j \epsilon_{ij} \chi_j(\textbf{x}_1)
\end{align}

or defining the **Fock operator** $f$,
\begin{align}
f(\textbf{x}_1) = h(\textbf{x}_1) + \sum_j J_j (\textbf{x}_1) - K_j (\textbf{x}_1) = h(\textbf{x}_1) + v^{HF}(\textbf{x}_1) \\
f(\textbf{x}_1) \chi_i(\textbf{x}_1) = \epsilon_i \chi_i (\textbf{x}_1)
\end{align}

The Hartree-Fock energy consists of three terms: (a) the one-electron energy which includes the kinetic energy of the electrons, and the attraction between the electrons and the nuclei; (b) the Coulomb energy which describes the interaction between an electron in spin-orbital $\chi_i$ and the *average* charge distribution of all the other electrons; and (c) the exchange energy which describes the exchange correlation between electrons with the same spin.

Clearly, the Hartree-Fock equation is an eigenvalue problem. Applying the Fock operator $f$ on spin-orbital $\chi_i$ will give the energy of the orbital, $\epsilon_i$, multiplied by the orbital $\chi_i$. However, notice that the Fock operator itself depends on the operator. So, the equation is usually solved iteratively. We start with a given set of orbitals and we solve the Hartree-Fock equation to get a new set of orbitals. Then, the calcuations are repeated until the solution converges. This is why the Hartree-Fock method is called a *self-consistent field* method.

The above Hartree-Fock equation is a complicated set of integro-differential equations that is difficult to solve numerically. Therefore, a more efficient procedure is required.

### The Orbital Energies and Koopman's Theorem
Again, the Hartree-Fock equation is given by:
\begin{align}
f(\textbf{x}_1) \chi_i(\textbf{x}_1) = \epsilon_i \chi_i (\textbf{x}_1)
\end{align}

Solving this equation yields the spin orbitals $\chi_i$ and the orbital energies $\epsilon_i$. As mentioned previously, some of these orbitals will be **occupied** and some of them will be **unoccupied** or **virtual**. We will label the occupied orbitals by $a, b, \cdots$ and the virutal orbitals by $r, s, \cdots$. What are the values for the occupied and virtual orbital energies $\epsilon_a$ and $\epsilon_r$?

Left-multiplying the above equation by $\chi_i (\textbf{x}_1)$, integrating, and using the definition of the Fock operator:
\begin{align}
\epsilon_i = [i|h|i] + \sum_{b=1}^N [ii|bb] - [ib|bi]
\end{align}

In particular, $\epsilon_a$ and $\epsilon_r$ become:
\begin{align}
\epsilon_a &= [a|h|a] + \sum_{b \neq a} [aa|bb] - [ab|ba] \\
\epsilon_r &= [r|h|r] + \sum_{b} [rr|bb] - [rb|br].
\end{align}

Now, what is the relationship between the orbital energies and the Hartree-Fock energy? If we add the orbtial energies we will get:
\begin{align}
\sum_a^N \epsilon_a = \sum_a^N [a|h|a] + \sum_a^N \sum_b^N [aa|bb] - [ab|ba].
\end{align}

However, the Hartree-Fock energy is given by:
\begin{align}
E_{HF} &= \sum_a^N [a|h|a] + \sum_{a \neq b}^N [aa|bb] - [ab|ba] \\
       &= \sum_a^N [a|h|a] + \frac{1}{2} \sum_a^N \sum_b^N [aa|bb] - [ab|ba].
\end{align}

Therefore, *the sum of the orbital energies is not equal to the Hartree-Fock energy*. The factor of 1/2 corrects for double-counting the interactions between the electrons in $\chi_a$ and $\chi_b$. So what is the physical meaning of the orbital energies $\epsilon_a$ and $\epsilon_r$?

From the above equation, we see that $\epsilon_a$ represents the energy of an electron in spin orbital $\chi_a$. The energy consists of the kinetic energy and attraction to the nuclei as well as a coulomb and exchange interactions with the remaining $N-1$ electrons. What is the physical meaning of $\epsilon_a$? Consider removing an electron from spin orbital $\chi_c$. What would be the energy of this $N-1$-electron system? Assuming the spin orbitals remain the same, the energy would be:

\begin{align}
{^{N-1}E_c} = \sum_{a \neq c}^N [a|h|a] + \frac{1}{2} \sum_{a \neq c}^N \sum_{b \neq c}^N [aa|bb] - [ab|ba].
\end{align}

The **ionization potential** is defined as the energy required to remove an electron from the system. Thus, the ionization potential is given by:
\begin{align}
\textrm{IP} &= {^{N-1}E_c} - {^{N}E_0} \\
            &= -[c|h|c] - \frac{1}{2} \sum_{a[b \equiv c]} [aa|bb] - [ab|ba] - \frac{1}{2} \sum_{b[a \equiv c]} [aa|bb] - [ab|ba] \\
            &= -[c|h|c] - \frac{1}{2} \sum_{a} [aa|cc] - [ac|ca] - \frac{1}{2} \sum_{b} [cc|bb] - [cb|bc] \\
            &= -[c|h|c] - \sum_{b} [cc|bb] - [cb|bc] \\
            &= -\epsilon_c
\end{align}

Thus, the energy of an occupied spin orbital represents the negative of the ionization potential, assuming the orbitals do not relax or change due to the removal of the electron.

Now consider the energy of a virtual spin orbital $\epsilon_r$. $\epsilon_r$ consists of the kinetic energy and attraction of an electron in the spin orbital $\chi_r$ to the nuclei, and also includes coulomb and exchange interactions with *all N electrons* of the Hartree-Fock ground state energy $|\Psi_0 \rangle$. In other words, $\epsilon_r$ represents the energy of an electron *added* to the system. The difference in energy between the $N-$electron and the $N+1$-electron systems is defined as the **electron affinity**. Assuming the orbitals do not change upon the addition of an electron, it can be shown that the electron affinity is equal to the negative of the virtual orbital energy
\begin{align}
\textrm{EA} = - \epsilon_r
\end{align}

This interpretation of the orbital energies as being the ionization potential and the electron affinity is called the **Koopman's theorem**. In practice, ionization potentials predicted using the Koopman's theorem are reasonably accurate, but the electron affinities are poor. The lack of accuracy arises from the neglect of orbital relaxation and electron correlation effects. Below we show some ionization potential data (in eV) computed using the Hartree-Fock theory with the aug-cc-pVQZ basis set. Again, we will see below how a Hartree-Fock calculation is actually carried out. The experimental data are taken from the [NIST](https://cccbdb.nist.gov/xp1x.asp?prop=8) website. In general, the Koopman's theorem results and the experimental data agree qualitatively.

|Element |Calculated |Experimental |
|--------|-----------|-------------|
|H       |  13.60    | 13.60       |
|He      |  24.98    | 24.59       |
|Li      |  5.34     | 5.39        |
|Be      |  8.42     | 9.32        |
|B       |  8.67     | 8.30        |
|C       |  11.95    | 11.26       |
|N       |  15.53    | 14.53       |
|O       |  16.64    | 13.62       |
|F       |  19.91    | 17.42       |
|Ne      |  23.15    | 21.56       |

In [58]:
# The orbital energies can be computed in Psi4 using the following syntax

import psi4
psi4.core.clean_options()

# Set psi4 output file
psi4.core.set_output_file('output.dat', True)

# Psi4 uses the density fitting approximation by default
# However, we want to use the exact two-electron integrals
# and therefore we will set the "scf_type" option.
# We also need to tell psi4 what Hartree-Fock flavor we want to use
psi4.set_options({"scf_type": "pk",
                  "reference": "uhf"})

# For defining the geometry, the first line defines the charge and multiplicity
psi4.geometry("""0 2
H
""")

e, wfn = psi4.energy("scf/Aug-cc-pVQZ", return_wfn=True)
# This extracts the energy for the highest occupied orbital
orbital_energy = wfn.epsilon_a_subset("AO", "OCC").np[-1]
orbital_energy_ev = orbital_energy * psi4.constants.hartree2ev
print("The energy of the highest occupied orbital is", orbital_energy_ev)

The energy of the highest occupied orbital is -13.604286765547206


### The Hartree-Fock-Roothan Equations

To solve the Hartree-Fock equations, Roothan suggested that we expand the spatial orbitals $\psi$ in terms of an atomic orbital basis, let's call them $\phi$. That is:
\begin{align}
\psi_i = \sum_{\mu=1}^K C_{\mu i} \phi_\mu
\end{align}

for each orbital $i$. If $K$ is infinite, we have a complete basis set and the expansion will be exact. However, in practice, a finite basis set is used. Substituting this expression in the Hartree-Fock equation and integrating out the spins (for restricted Hartree-Fock theory), we get:
\begin{align}
f(\textbf{r}_1) \sum_\nu C_{\nu i} \phi_\nu (\textbf{r}_1) = 
 \epsilon_i \sum_\nu C_{\nu i} \phi_\nu (\textbf{r}_1)).
\end{align}

If we left-multiply by $\phi_\mu^*$ and integrate we get the matrix equation:
\begin{align}
\sum_\nu F_{\mu \nu} C_{\nu i} &= \epsilon_i \sum_\nu S_{\mu \nu} C_{\nu i} \\
\textbf{FC} &=\textbf{SC} \epsilon,
\end{align}
where
\begin{align}
S_{\mu \nu} &= \int d \textbf{r}_1 \phi_\mu^* (\textbf{r}_1) \phi_\nu(\textbf{r}_1) \\
F_{\mu \nu} &= \int d \textbf{r}_1 \phi_\mu^* f(\textbf{r}_1) \phi_\nu(\textbf{r}_1).
\end{align}

This is a matrix equation. $\textbf{S}$ is called the **overlap matrix**. It is needed because we do not assume in general that the basis functions $\phi_i$ are orthogonal. $\textbf{F}$ is the Fock matrix. $\textbf{C}$ is the orbital coefficient matrix, where each column represents a molecular orbital $\psi$. Finally, $\epsilon$ is a diagonal matrix of the orbital energies. Thus, expanding the orbitals $\psi$ in terms of the basis functions $\phi$ converts the Hartree-Fock equation to a matrix equation that can be solved by standard techniques of linear algebra.

The Hartree-Fock-Roothan equation is called a pseduo-eigenvalue equation. Solving the equation, we get the orbital coefficients $\textbf{C}$ and the orbital energies $\epsilon$. However, notice again that the Fock matrix depends on the orbital. So again we need to solve this equation in an iterative manner. We start with a guess $\textbf{C}$, then we solve the equation to get a new value for $\textbf{C}$. We repeat the calculation until our orbital coefficients and our orbital energies converge. This is called a *self-consistent field* procedure.

### Forming the Fock Matrix (RHF)
The Fock operator in restricted Hartree-Fock theory is given by:
\begin{align}
f(\textbf{r}) = h(\textbf{r}) + \sum_a^{N/2} 2 J_a (\textbf{r}) - K_a (\textbf{r})
\end{align}

The Fock matrix elements in the $\{\phi\}$ basis become:
\begin{align}
F_{\mu \nu} &= \int d \textbf{r}_1 \phi_\mu^* f(\textbf{r}_1) \phi_\nu(\textbf{r}_1) \\
            &= H_{\mu \nu} + \sum_a^{N/2} 2 (\mu \nu|a a) - (\mu a|a\nu).
\end{align}
$H$ is called the **core Hamiltonian** matrix and it does not depend on the density.

If we also express the coulomb and exchange integrals in terms of the $\{\phi\}$ basis we get:
\begin{align}
F_{\mu \nu} &= H_{\mu \nu} + \sum_a^{N/2} \sum_{\lambda \sigma}^K C_{\lambda a}^* C_{\sigma a} \left ( 2 (\mu \nu | \lambda \sigma) - (\mu \sigma | \lambda \nu) \right ) \\
            &= H_{\mu \nu} + \sum_{\lambda \sigma}^K D_{\sigma \lambda} \left ( (\mu \nu | \lambda \sigma) - \frac{1}{2}(\mu \sigma | \lambda \nu) \right ).
\end{align}

$D_{\sigma \lambda}$ is called the **density matrix** and it is given by:
\begin{align}
D_{\sigma \lambda} = \sum_a^{N/2} 2 C_{\lambda a}^* C_{\sigma a}.
\end{align}

The calculation of the four-center, two-electron integrals are the most expensive step in the Hartree-Fock procedure. That is, it will take the longest time in the Hartree-Fock calculation. Hartree-Fock calculations are said to scale as $O(n^4)$, where $n$ is the number of basis functions. For example, doubling the number of basis functions will increase the cost of a Hartree-Fock calculation 16 times. There are various approaches to lower the computational cost of the Hartree-Fock procedure.

### Summary of the Self-Consistent Field Procedure
1. Specify molecule, basis function, and spin state (multiplicity).
2. Form the overlab matrix $\textbf{S}$ from the basis functions.
3. Provide an initial guess for the orbitals $\textbf{C}$.
4. Calculate the Fock matrix $\textbf{F}$.
5. Solve the Hartree-Fock-Roothan equation: $\textbf{FC} =\textbf{SC} \epsilon$ to get a new orbital matrix $\textbf{C}$.
6. Calculate the Fock matrix $\textbf{F}$ use the new orbital matrix $\textbf{C}$.
7. Repeat steps 5 and 6 until the orbital matrix $\textbf{C}$ and the orbital energies

### Practical points
#### Initial Guess
Since the SCF procedue is iterative, a good initial guess can accelerate the convergence. There are various ways to guess the initial density or orbital coefficient matrix. The easiest way is to set the matrix to zero. This leads to the Fock matrix being equal to the core Hamiltonian. This is the **core** guess. This is usually a poor choice because it completely neglects electron-electron interactions. Other guesses include the **extended Huckel** guess and the **superposition of atomic densities (SAD)**, among others. It is also often fruitful to **read** the molecular orbitals from a previously converged calculations on a similar system.

#### Accelerating Convergence
Often, the SCF procedue will not converge or will converge slowly. Useful methods to accelerate the convergence are direct inversion of the iterative subspace (**DIIS**), which is an extrapolation technique. Other useful methods are the maximum overlap method (**MOM**), and **damping**, which are particularly useful when the SCF procedure is oscillating. 

#### Lowering the Scaling of the Hartree-Fock Method
As mentioned above, the Hartree-Fock method scaled as $O(n^4)$. This can be lowered to $O(n^3)$ using techniques like **density-fitting** or **resolution-of-the-identity**. Furthermore, screening techniques can be used to neglect the interactions between basis functions that are far from each other, thus lowering the computational cost to $O(n^2)$.

#### What Can We Calculate with Hartree-Fock Theory?
- Hartree-Fock energy
- Equilibrium geometry
- Ionization potential
- Electrostatic potential
- Dipole moment
- Molecular orbitals
- ...

### What is next?
In practice, Hartree-Fock calculations, even at the complete basis set limit, are not very accurate because of the neglect of electron correlation. The total energy is defined as the sum of the Hartree-Fock energy at the complete basis set limit and the energy of electron correlation:
\begin{align}
E_0 = E_{HF} + E_{corr}
\end{align}

Calculating accurate $E_{corr}$ is required for chemical applications. Therefore, many theoretical methods have been developed for its computation. Most of these methods build on the results of Hartree-Fock theory and therefore are called **post-Hartree-Fock** methods. These methods are generally more accurate than the Hartree-Fock method but they are computationally expensive. These methods, including Hartree-Fock, are called *ab initio* (from first principle) because they only depend on numerical approximations without fitting to experimental data.

On the other hand, **semi-empirical** methods seek to reduce the computational cost of the Hartree-Fock method by neglecting some 2-electron integrals and replacing them with parameters fitted to experimental data. This extends their applicability to large systems, but makes them dependent on the reliability of the fitting procedure.

## Useful Resources

- Szabo, A.; Ostlund, N. S. *Modern Quantum Chemistry: Introduction to Advanced Electronic Structure Theory*; Dover Publications: Mineola, NY, 1996. (Chapters 2 and 3)  
- Cramer, C. J. *Essentials of Computational Chemistry: Theories and Models*, 2nd ed.; John Wiley & Sons: Chichester, England, 2004. (Chapters 4 and 6)
- Jensen, F. *Introduction to Computational Chemistry*, 3rd ed.; John Wiley & Sons: Nashville, TN, 2017. (Chapter 3)
- Sherrill's notes: http://vergil.chemistry.gatech.edu/courses/chem6485/pdf/Hartree-Fock-Intro.pdf; http://vergil.chemistry.gatech.edu/notes/hf-intro/hf-intro.pdf